Reporting changes

jorblancoa opened the pull request:
- Refactor report module
- Select report type at runtime (SONATA or Bin)
- Adapt integration tests

ohm314 commented on coreneuron/utils/reports/sonata_report_handler.cpp:
```diff
@@ -0,0 +1,49 @@
+    for(const auto& kv: vars_to_report) {
+        int gid = kv.first;
+        const std::vector<VarWithMapping> &vars = kv.second;
+        if (!vars.size())
```
use `vars.empty()`

ohm314 replied:
no? didn't like that? :)

ohm314 commented on extra/nrnivmodl-core.in:
```diff
@@ -12,6 +12,16 @@ MAKE_OPTIONS="MECH_NAME NMODL_BINARY DESTDIR INCFLAGS LINKFLAGS MODS_PATH VERBOS
 PARALLELISM=4
 SAVE_FILE="@CMAKE_HOST_SYSTEM_PROCESSOR@/nrnivmodl_options.txt"
 
+# To support build + install later we save/restore options
```
is that not a change that is already on master?

jorblancoa replied:
Yes, but i needed it for testing with spack so I cherrypicked it from the branch before it was merged

tristan0x commented on coreneuron/nrniv/main1.cpp:
```diff
@@ -518,7 +522,14 @@ extern "C" int run_solve_core(int argc, char** argv) {
         for (size_t i = 0; i < configs.size(); i++) {
-            register_report(dt, tstop, delay, configs[i]);
+            std::shared_ptr<ReportHandler> report_handler;
+            if(0 == strcmp( configs[i].format, "Bin")) {
```
You should use `std::strcmp` and include `<cstring>` header instead of `<string.h>`

ohm314 replied:
Good catch! that one had actually bothered me too :)

tristan0x commented on coreneuron/utils/reports/binary_report_handler.cpp:
```diff
@@ -0,0 +1,129 @@
+        std::vector<VarWithMapping>& vars = it->second;
+        if (!vars.size())
+            continue;
+        NrnThreadMappingInfo* mapinfo = (NrnThreadMappingInfo*)nt.mapping;
```
Indeed, the proper cast should be used instead of a C-style cast, something like:
```cpp
        const auto* mapinfo = static_cast<NrnThreadMappingInfo*>(nt.mapping);
```
(Maybe `const` qualifier needs to be removed).

tristan0x commented on coreneuron/utils/reports/binary_report_handler.cpp:
```diff
@@ -0,0 +1,129 @@
+        int gid = it->first;
+        std::vector<VarWithMapping>& vars = it->second;
+        if (!vars.size())
+            continue;
```
Single statement for conditions and loop constructions should be put into braces, for safety purposes.

tristan0x commented on coreneuron/utils/reports/binary_report_handler.cpp:
```diff
@@ -0,0 +1,129 @@
+        extra[1] = m->get_seclist_segment_count("soma");
+
+        /** for this gid, get mapping information */
+        records_add_report((char*)config.output_path, gid, gid, gid, config.start, config.stop,
```
Please avoid C-Style cast

tristan0x commented on coreneuron/utils/reports/binary_report_handler.cpp:
```diff
@@ -0,0 +1,129 @@
+        records_set_report_max_buffer_size_hint((char*)config.output_path, config.buffer_size);
+        /** add extra mapping */
+        records_extra_mapping(config.output_path, gid, 5, extra);
+        for (int var_idx = 0; var_idx < vars.size(); ++var_idx) {
```
even `for (const auto& var: vars)` I guess

tristan0x commented on coreneuron/utils/reports/binary_report_handler.cpp:
```diff
@@ -0,0 +1,129 @@
+    int sizemapping = 1;
+    int extramapping = 5;
+    int mapping[] = {0};
+    int extra[] = {1, 0, 0, 0, 1};
```
I would use a `std::array` to have compile time errors in case of writing issue.

tristan0x commented on coreneuron/utils/reports/binary_report_handler.cpp:
```diff
@@ -0,0 +1,129 @@
+        extra[2] = m->get_seclist_section_count("axon");
+        extra[3] = m->get_seclist_section_count("dend");
+        extra[4] = m->get_seclist_section_count("apic");
+        extra[0] = extra[1] + extra[2] + extra[3] + extra[4];
```
Assuming `extra` is a `std::array<int, 5>`, I would use `std::accumulate(extra.begin() + 1, extra.end(), 0)` so that it won't be required to update this instruction next time a new field needs to be reported.

tristan0x commented on coreneuron/utils/reports/binary_report_handler.cpp:
```diff
@@ -0,0 +1,129 @@
+        extra[4] = m->get_seclist_section_count("apic");
+        extra[0] = extra[1] + extra[2] + extra[3] + extra[4];
+
+        records_add_report((char*)config.output_path, gid, gid, gid, config.start, config.stop,
```
Please avoid C-style cast.

tristan0x commented on coreneuron/utils/reports/binary_report_handler.cpp:
```diff
@@ -0,0 +1,129 @@
+        records_set_report_max_buffer_size_hint((char*)config.output_path, config.buffer_size);
+        /** add extra mapping */
+        records_extra_mapping(config.output_path, gid, 5, extra);
+        for (int var_idx = 0; var_idx < vars.size(); ++var_idx) {
```
Like @ohm314 said, you can use range-based loop here.

tristan0x commented on coreneuron/utils/reports/binary_report_handler.cpp:
```diff
@@ -0,0 +1,129 @@
+    int sizemapping = 1;
+    int extramapping = 5;
+    int mapping[] = {0};
+    int extra[] = {1, 0, 0, 0, 1};
```
Prefer `std::array<int, 5>` here.

tristan0x commented on coreneuron/utils/reports/report_event.cpp:
```diff
@@ -0,0 +1,43 @@
+}
+
+/** on deliver, call ReportingLib and setup next event */
+void ReportEvent::deliver(double t, NetCvode *nc, NrnThread *nt) {
```
If it doesn't break the API, `nc` and `nt` should be references instead of pointers if there are never `nullptr`.

tristan0x commented on coreneuron/utils/reports/report_event.cpp:
```diff
@@ -0,0 +1,43 @@
+#pragma omp critical
+    {
+        // each thread needs to know its own step
+        sonata_record_node_data(step, gids_to_report.size(), &gids_to_report[0], report_path);
```
```suggestion
        sonata_record_node_data(step, gids_to_report.size(), gids_to_report.data(), report_path);
```
??

tristan0x commented on coreneuron/utils/reports/report_event.cpp:
```diff
@@ -0,0 +1,43 @@
+    {
+        // each thread needs to know its own step
+        sonata_record_node_data(step, gids_to_report.size(), &gids_to_report[0], report_path);
+        records_nrec(step, gids_to_report.size(), &gids_to_report[0], report_path);
```
```suggestion
        records_nrec(step, gids_to_report.size(), gids_to_report.data(), report_path);
```

tristan0x commented on coreneuron/utils/reports/report_event.h:
```diff
@@ -0,0 +1,40 @@
+};
+
+// mapping the set of variables pointers to report to its gid
+typedef std::map<int, std::vector<VarWithMapping> > VarsToReport;
```
Use `unordered_map` if ordering during traversal is not required.

tristan0x commented on coreneuron/utils/reports/report_event.h:
```diff
@@ -0,0 +1,40 @@
+#ifdef ENABLE_REPORTING
+struct VarWithMapping {
+    int id;
+    double *var_value;
```
Maybe consider `gsl::span` here?

tristan0x commented on coreneuron/utils/reports/report_event.h:
```diff
@@ -0,0 +1,40 @@
+
+class ReportEvent : public DiscreteEvent {
+  public:
+    ReportEvent(double dt, double tstart, VarsToReport &filtered_gids, const char *name);
```
IMO `filtered_gids` should be marked const.

tristan0x commented on coreneuron/utils/reports/report_handler.cpp:
```diff
@@ -0,0 +1,198 @@
+    m_report_config.mech_id = nrn_get_mechtype(m_report_config.mech_name);
+    if (m_report_config.type == SynapseReport && m_report_config.mech_id == -1) {
+        std::cerr << "[ERROR] mechanism to report: " << m_report_config.mech_name
+                  << " is not mapped in this simulation, cannot report on it" << std::endl;
```
Do not unnecessarily flush the stream.
```suggestion
                  << " is not mapped in this simulation, cannot report on it\n";
```

tristan0x commented on coreneuron/utils/reports/report_handler.cpp:
```diff
@@ -0,0 +1,198 @@
+    }
+    for (int ith = 0; ith < nrn_nthread; ++ith) {
+        NrnThread &nt = nrn_threads[ith];
+        if (!nt.ncell)
```
Please protect the statement around braces.

tristan0x commented on coreneuron/utils/reports/report_handler.cpp:
```diff
@@ -0,0 +1,198 @@
+        NrnThread &nt = nrn_threads[ith];
+        if (!nt.ncell)
+            continue;
+        std::vector<int> nodes_to_gid = map_gids(nt);
```
```suggestion
        const std::vector<int>& nodes_to_gid = map_gids(nt);
```
?

tristan0x commented on coreneuron/utils/reports/report_handler.cpp:
```diff
@@ -0,0 +1,198 @@
+    }
+#else
+    if (nrnmpi_myid == 0) {
+        printf("\n WARNING! : Can't enable reports, recompile with ReportingLib! \n");
```
Use `std::cout` instead?

tristan0x commented on coreneuron/utils/reports/report_handler.cpp:
```diff
@@ -0,0 +1,198 @@
+    NrnThreadMappingInfo *mapinfo = (NrnThreadMappingInfo *) nt.mapping;
+
+    if (!mapinfo) {
+        std::cout << "[SOMA] Error : mapping information is missing for a Cell group " << std::endl;
```
```suggestion
        std::cerr << "[SOMA] Error : mapping information is missing for a Cell group\n";
```

tristan0x commented on coreneuron/utils/reports/report_handler.cpp:
```diff
@@ -0,0 +1,198 @@
+}
+
+VarsToReport
+ReportHandler::get_compartment_vars_to_report(NrnThread &nt, std::set<int> &target, double *report_variable) {
```
* I guess `target` can be const here
* consider gsl::span for report_variable

tristan0x commented on coreneuron/utils/reports/report_handler.cpp:
```diff
@@ -0,0 +1,198 @@
+    VarsToReport vars_to_report;
+    NrnThreadMappingInfo *mapinfo = (NrnThreadMappingInfo *) nt.mapping;
+    if (!mapinfo) {
+        std::cout << "[COMPARTMENTS] Error : mapping information is missing for a Cell group "
```
Prefer std::cerr for consistency

tristan0x commented on coreneuron/utils/reports/sonata_report_handler.h:
```diff
@@ -0,0 +1,24 @@
+class SonataReportHandler: public ReportHandler {
+  public:
+    SonataReportHandler(ReportConfiguration& config): ReportHandler(config) {}
+    ~SonataReportHandler() = default;
```
Is the default destructor required here?

tristan0x commented on tests/integration/integration_test.sh.in:
```diff
@@ -22,6 +22,26 @@ fi
+
+  if [ ! -f out_SONATA.dat ]
+  then
+    echo "No SONATA output files. Test failed!"
```
```suggestion
    echo "[ERROR] No SONATA output files. Test failed!" >&2
```

tristan0x commented on tests/integration/integration_test.sh.in:
```diff
@@ -22,6 +22,26 @@ fi
+  diff -w out_SONATA.dat @CMAKE_CURRENT_SOURCE_DIR@/@SIM_NAME@/out.dat.ref > diff_SONATA.dat 2>&1
+  if [ -s diff_SONATA.dat ]
+  then
+    echo "SONATA Results are different, check the file diff_SONATA.dat. Test failed!"
```
```suggestion
    echo "[ERROR] SONATA Results are different, check the file diff_SONATA.dat. Test failed!" >&2
```

tristan0x commented on tests/integration/negative_integration_test.sh.in:
```diff
@@ -24,6 +24,26 @@ fi
+
+  if [ ! -f out_SONATA.dat ]
+  then
+    echo "No SONATA output files. Test failed!"
```
```suggestion
    echo "[ERROR] No SONATA output files. Test failed!" >&2
```

tristan0x commented on tests/integration/negative_integration_test.sh.in:
```diff
@@ -24,6 +24,26 @@ fi
+  diff -w out_SONATA.dat @CMAKE_CURRENT_SOURCE_DIR@/@SIM_NAME@/out.dat.ref > diff_SONATA.dat 2>&1
+  if [ ! -s diff_SONATA.dat ]
+  then
+    echo "Results are identical for SONATA file when they should not, check the file out_SONATA.dat. Test failed!"
```
```suggestion
    echo "[ERROR] Results are identical for SONATA file when they should not, check the file out_SONATA.dat. Test failed!" >&2
```

tristan0x commented on tests/integration/reportinglib/1.check.in:
```diff
@@ -5,12 +5,14 @@
+   if [ $nb_recorded -eq 100 ] && [ $nb_recorded_sonata -eq 101 ]; then
      exit $OK
    else
      exit $FAILED
```
Maybe an error message would help L18

jorblancoa commented on coreneuron/utils/reports/binary_report_handler.cpp:
```diff
@@ -0,0 +1,129 @@
+    int extra[] = {1, 0, 0, 0, 0};
+
+    VarsToReport::iterator it;
+    for (it = vars_to_report.begin(); it != vars_to_report.end(); ++it) {
```
Yes! this was how was the old code, i just restructured the code to introduce the new reports.
But i guess it makes sense to change this things since are really quickly!

jorblancoa commented on coreneuron/utils/reports/report_event.cpp:
```diff
@@ -0,0 +1,43 @@
+    gids_to_report.reserve(filtered_gids.size());
+    for (it = filtered_gids.begin(); it != filtered_gids.end(); ++it) {
+        gids_to_report.push_back(it->first);
+    }
```
I think its not really clear.
Maybe change it to:
`for(const auto& gid: filtered_gids) {
        gids_to_report.push_back(gid.first);
}`

jorblancoa commented:
Thanks for your review!! 
To be honest most of the comments were about the old reportinglib. I just introduced the new SONATA reports and restructured the classes a bit, I didn't plan to change the old reportinglib code since it will be deleted soon.

However I addressed most of the comments since most of them were quick fixes. Thanks again!

jorblancoa commented on coreneuron/utils/reports/report_event.h:
```diff
@@ -0,0 +1,41 @@
+#include "coreneuron/nrniv/netcvode.h"
+
+#define MAX_REPORT_NAME_LEN 256
+#define MAX_FILEPATH_LEN 4096
```
You are right they are defined already in nrnreport.h
maybe rename to:
REPORT_MAX_NAME_LEN
REPORT_MAX_FILEPATH_LEN

tristan0x replied:
Can't `report_event.h` include `nrnreport.h` ?

jorblancoa replied:
Yes, actually report_event.h doesn't need the constants since the char report_path[MAX_FILEPATH_LEN] used can be changed easily to std::string

tristan0x replied:
Even better!

ohm314 commented:
@jorblancoa : is this ready for merging?
@tristan0x are you happy with the PR after the latest changes?

jorblancoa commented:
> @jorblancoa : is this ready for merging?

Yes!

tristan0x commented on CMakeLists.txt:
```diff
@@ -147,19 +146,29 @@ if(NOT CORENRN_ENABLE_TIMEOUT)
+  if(NOT reportinglib_FOUND AND NOT sonata_FOUND)
+    message(FATAL_ERROR "NO reporting library was found")
+  elseif(reportinglib_FOUND AND sonata_FOUND)
+    add_definitions("-DENABLE_REPORTINGLIB")
```
Macros should be prefixed by their project. Something like `CORENRN_ENABLE_SONATA_REPORTS`

jorblancoa replied:
I didn't add CORENRN_ prefix to the macro names because none of the other macros have the prefix, but I like the change REPORTINGLIB for BIN_REPORTS

pramodk replied:
fair! we can do that change separately. By the way there is a separate PR #167 but that needs some extra work. Let's take this separately.

tristan0x commented on CMakeLists.txt:
```diff
@@ -147,19 +146,29 @@ if(NOT CORENRN_ENABLE_TIMEOUT)
+      set(reportinglib_INCLUDE_DIR "")
+      set(reportinglib_LIBRARY "")
+    else()
+      message(FATAL_ERROR "SONATA library was found without reports")
```
```suggestion
      message(SEND_ERROR "SONATA library was found but without reporting support")
```

tristan0x commented on coreneuron/apps/main1.cpp:
```diff
@@ -518,7 +523,14 @@ extern "C" int run_solve_core(int argc, char** argv) {
+                report_handler = std::make_unique<BinaryReportHandler>(configs[i]);
+            } else if (std::strcmp(configs[i].format, "SONATA") == 0) {
+                report_handler = std::make_unique<SonataReportHandler>(configs[i]);
+            }
```
An `else` clause is missing. It should report an error that format is neither "Bin"  nor "SONATA".
Besides, I would move the creation of a `ReportHandler` from a `ReportConfiguration` in a dedicated function.

tristan0x commented on coreneuron/io/reports/binary_report_handler.cpp:
```diff
@@ -0,0 +1,95 @@
+}
+
+static void create_compartment_extra(const CellMapping& mapping, std::array<int, 5>& extra) {
+    extra = {1, 0, 0, 0, 1};
```
Instruction L23 is not necessary.

tristan0x commented on coreneuron/io/reports/binary_report_handler.cpp:
```diff
@@ -0,0 +1,95 @@
+static void create_custom_extra(const CellMapping& mapping, std::array<int, 5>& extra) {
+    extra = {1, 0, 0, 0, 1};
+    extra[1] = mapping.get_seclist_section_count("soma");
+    // todo: axon seems masked on neurodamus side, need to check
```
There is a _todo_

tristan0x commented on coreneuron/io/reports/binary_report_handler.cpp:
```diff
@@ -0,0 +1,95 @@
+                                          create_extra_func& create_extra) {
+    int sizemapping = 1;
+    int extramapping = 5;
+    std::array<int, 1> mapping = {0};
```
what about `int mapping{}` instead?

tristan0x commented on coreneuron/io/reports/report_event.cpp:
```diff
@@ -0,0 +1,51 @@
+}
+
+/** on deliver, call ReportingLib and setup next event */
+void ReportEvent::deliver(double t, NetCvode* nc, NrnThread* nt) {
```
Why not using a reference here? Can `nc` and `nt` be `nullptr`?

pramodk replied:
This is derived from base class called DiscreteEvent and hence ...

tristan0x commented on coreneuron/io/reports/report_event.hpp:
```diff
@@ -0,0 +1,38 @@
+};
+
+// mapping the set of variables pointers to report to its gid
+typedef std::map<int, std::vector<VarWithMapping> > VarsToReport;
```
isn't a `std::unordered_map` enough?
Besides I would use the `using` keyword to declare `VarsToReport`. Something like:

```suggestion
using VarsToReport =std::unordered_map<int, std::vector<VarWithMapping>>;
```

tristan0x commented on coreneuron/io/reports/report_event.hpp:
```diff
@@ -0,0 +1,38 @@
+    ReportEvent(double dt, double tstart, const VarsToReport& filtered_gids, const char* name);
+
+    /** on deliver, call ReportingLib and setup next event */
+    virtual void deliver(double t, NetCvode* nc, NrnThread* nt);
```
```suggestion
    void deliver(double t, NetCvode* nc, NrnThread* nt) override;
```

tristan0x commented on coreneuron/io/reports/report_event.hpp:
```diff
@@ -0,0 +1,38 @@
+
+    /** on deliver, call ReportingLib and setup next event */
+    virtual void deliver(double t, NetCvode* nc, NrnThread* nt);
+    virtual bool require_checkpoint();
```
```suggestion
    bool require_checkpoint() override;
```

Besides, `DiscreteEvent::require_checkpoint` should be marked with `const` specifier.

tristan0x commented on coreneuron/io/reports/report_handler.cpp:
```diff
@@ -0,0 +1,230 @@
+                register_custom_report(nt, m_report_config, vars_to_report);
+        }
+        if (!vars_to_report.empty()) {
+            std::unique_ptr<ReportEvent> report_event =
```
```suggestion
            auto report_event =
```

tristan0x commented on coreneuron/io/reports/report_handler.cpp:
```diff
@@ -0,0 +1,230 @@
+    }
+#else
+    if (nrnmpi_myid == 0) {
+        std::cerr << "[WARNING] : Can't enable reports, recompile with ReportingLib! \n";
```
```suggestion
        std::cerr << "[WARNING] : reporting is disabled. Please recompile with either SONATA or ReportingLib\n";
```

tristan0x commented on coreneuron/io/reports/report_handler.cpp:
```diff
@@ -0,0 +1,230 @@
+                                         ReportConfiguration& config,
+                                         const VarsToReport& vars_to_report) {
+    if (nrnmpi_myid == 0) {
+        std::cerr << "[WARNING] : Report '" << config.output_path << "' not supported!\n";
```
I don't know what it means when `nrnmpi_myid == 0`, but the error message does not provide much more information. It should be improved if there is a more precise reason.

Besides if the exclamation mark at the end of the logs messages is uncommon in CoreNeuron, then IMO they should be removed.

pramodk replied:
Same question and I couldn't figure out  why all reports here print the warning with `config.output_path` ?

jorblancoa replied:
This is called only when the format of the report doesn't exist. Otherwise the subclasses sonata_report_handler or binary_report_handler are called.

tristan0x commented on coreneuron/io/reports/report_handler.cpp:
```diff
@@ -0,0 +1,230 @@
+    const auto* mapinfo = static_cast<NrnThreadMappingInfo*>(nt.mapping);
+    if (!mapinfo) {
+        std::cerr << "[SOMA] Error : mapping information is missing for a Cell group " << nt.ncell
+                  << "\n";
```
```suggestion
                  << '\n';
```
I know the process is about to exit, but still :)

tristan0x commented on coreneuron/io/reports/report_handler.cpp:
```diff
@@ -0,0 +1,230 @@
+            CellMapping* m = mapinfo->get_cell_mapping(gid);
+            if (m == nullptr) {
+                std::cerr << "[SOMA] Error : Soma mapping information is missing for gid " << gid
+                          << "\n";
```
```suggestion
                          << '\n';
```

tristan0x commented on coreneuron/io/reports/report_handler.cpp:
```diff
@@ -0,0 +1,230 @@
+    const auto* mapinfo = static_cast<NrnThreadMappingInfo*>(nt.mapping);
+    if (!mapinfo) {
+        std::cerr << "[COMPARTMENTS] Error : mapping information is missing for a Cell group "
+                  << nt.ncell << "\n";
```
```suggestion
                  << nt.ncell << '\n';
```

tristan0x commented on coreneuron/io/reports/report_handler.cpp:
```diff
@@ -0,0 +1,230 @@
+            if (cell_mapping == nullptr) {
+                std::cerr
+                    << "[COMPARTMENTS] Error : Compartment mapping information is missing for gid "
+                    << gid << "\n";
```
```suggestion
                    << gid << '\n';
```

tristan0x commented on tests/integration/negative_integration_test.sh.in:
```diff
@@ -24,9 +24,29 @@ fi
+reports=@ENABLE_SONATA_REPORTS@
+if [ "$reports" = "ON" ]
+then
+  data=$(h5dump -d /spikes/timestamps -d /spikes/node_ids -y -O spikes | tr "," "\n")
```
`h5dump ...` command exit status is not checked. It would be nice to emit an error and prevent further processing if it fails.

tristan0x commented on tests/integration/negative_integration_test.sh.in:
```diff
@@ -24,9 +24,29 @@ fi
+
+  if [ ! -f out_SONATA.dat ]
+  then
+    echo "[ERROR] No SONATA output files. Test failed!" >&2
```
```suggestion
    echo "[ERROR] expected SONATA output file 'out_SONATA.dat' is missing. Test failed!" >&2
```

tristan0x commented on tests/integration/reportinglib/1.check.in:
```diff
@@ -4,18 +4,45 @@
+  then
+     #checking something with @reportinglib_BIN_DIR@/dump test_1.bbp
+     # for ascii files that will be using grep for instance.
+     nb_recorded=$(@reportinglib_LIB_DIR@/../usr/local/bin/somaDump test_1.bbp 1| wc -l)
```
You should make sure that `somaDump` executable exists before calling it. Actually, it should be done by CMake and have a `@SomaDump_EXECUTABLE@` variable

tristan0x commented on tests/integration/reportinglib/1.check.in:
```diff
@@ -4,18 +4,45 @@
+     if [ $nb_recorded -eq 100 ]; then
+       exit $OK
+     else
+       echo "[ERROR] The number of steps recorded doesn't match. Test failed!" >&2
```
```suggestion
       echo "[ERROR] The number of steps recorded by ReportingLib doesn't match. Found $nb_recorded, expected 100. Test failed!" >&2
```

tristan0x commented on tests/integration/reportinglib/1.check.in:
```diff
@@ -4,18 +4,45 @@
+       exit $FAILED
+     fi
+  else
+     echo "[ERROR] reportinglib soma file doesn't exist. Test failed!" >&2
```
```suggestion
     echo "[ERROR] expected ReportingLib soma file 'test_1.bbp' is missing. Test failed!" >&2
```

tristan0x commented on tests/integration/reportinglib/1.check.in:
```diff
@@ -4,18 +4,45 @@
+  then
+     #checking something with @reportinglib_BIN_DIR@/dump test_1.bbp
+     # for ascii files that will be using grep for instance.
+     nb_recorded_sonata=$(h5dump -d /report/data -y -O test_2.h5 | wc -l)
```
Are you sure `h5dump` is available in PATH? It should be found by CMake and use in this script with a variable like `@H5DUMP_EXECUTABLE@`

tristan0x commented on tests/integration/reportinglib/1.check.in:
```diff
@@ -4,18 +4,45 @@
+     if [ $nb_recorded_sonata -eq 101 ]; then
+       exit $OK
+     else
+       echo "[ERROR] The number of steps recorded doesn't match. Test failed!" >&2
```
```suggestion
       echo "[ERROR] The number of steps recorded by SONATA doesn't match.  Found $nb_recorded, expected 101. Test failed!" >&2
```

tristan0x commented on tests/integration/reportinglib/1.check.in:
```diff
@@ -4,18 +4,45 @@
+       exit $FAILED
+     fi
+  else
+     echo "[ERROR] SONATA reports soma file doesn't exist. Test failed!" >&2
```
```suggestion
     echo "[ERROR] expected SONATA reports soma file 'test_2.h5` doesn't exist. Test failed!" >&2
```

ohm314 commented:
> Minor remarks. Otherwise LGTM.
> Besides it is my understanding that clang-format has not been applied to this patch. It should be the case.

@tristan0x : `clang-format` will be applied in one of the next PRs that will bring `hpc-coding-conventions`.

pramodk commented on coreneuron/io/output_spikes.cpp:
```diff
@@ -43,8 +43,8 @@ THE POSSIBILITY OF SUCH DAMAGE.
 #ifdef ENABLE_SONATA_REPORTS
-#include "reportinglib/records.h"
-#endif
+#include "bbp/sonata/reports.h"
```
just curious : is there `bbp` in the include path? surprised by this!

jorblancoa replied:
Yes, libsonata directory structure for include is the following:
include/bbp/sonata/*.h

pramodk commented on coreneuron/io/reports/nrnreport.cpp:
```diff
@@ -0,0 +1,105 @@
+/*
+Copyright (c) 2016, Blue Brain Project
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without modification,
+are permitted provided that the following conditions are met:
+1. Redistributions of source code must retain the above copyright notice,
+   this list of conditions and the following disclaimer.
+2. Redistributions in binary form must reproduce the above copyright notice,
+   this list of conditions and the following disclaimer in the documentation
+   and/or other materials provided with the distribution.
+3. Neither the name of the copyright holder nor the names of its contributors
+   may be used to endorse or promote products derived from this software
+   without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+THE POSSIBILITY OF SUCH DAMAGE.
+*/
```
```suggestion
```

pramodk commented on coreneuron/io/reports/report_event.hpp:
```diff
@@ -0,0 +1,39 @@
+#include <algorithm>
+#include <unordered_map>
+#include <vector>
+#include <string>
```
```suggestion
#include <string>

```

jorblancoa commented on coreneuron/io/output_spikes.cpp:
```diff
@@ -165,8 +165,10 @@ void output_spikes_parallel(const char* outpath) {
 #ifdef ENABLE_SONATA_REPORTS
-    records_write_spikes(spikevec_time, spikevec_gid);
-#else
+    const char* population_name = "All";
```
True, we said we should add at some point the population as an input.
Will put a TODO in the meantime

jorblancoa commented on tests/unit/sdprintf/CMakeLists.txt:
```diff
@@ -0,0 +1,18 @@
+# =============================================================================
```
the file? or the copyright?
